pkg/services/ngalert/notifier/channels/email.go

```diff
@@ -114,13 +117,48 @@ func (en *EmailNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool,
 				"ExternalURL":       data.ExternalURL,
 				"RuleUrl":           ruleURL,
 				"AlertPageUrl":      alertPageURL,
+				"ImageLink":         "",
+				"EmbeddedImage":     "",
 			},
 			To:          en.Addresses,
 			SingleEmail: en.SingleEmail,
 			Template:    "ng_alert_notification",
 		},
 	}
 
+	// TODO: modify the email sender code to support multiple file or image URL
+	// fields. We cannot use images from every alert yet.
+	imgToken := getTokenFromAnnotations(as[0].Annotations)
+	if len(imgToken) != 0 {
+		timeoutCtx, cancel := context.WithTimeout(ctx, ImageStoreTimeout)
+		imgURL, err := en.images.GetURL(timeoutCtx, imgToken)
+		cancel()
+		if err != nil {
+			if !errors.Is(err, ErrImagesUnavailable) {
+				// Ignore errors. Don't log "ImageUnavailable", which means the storage doesn't exist.
+				en.log.Warn("failed to retrieve image url from store", "error", err)
+			}
+		} else if len(imgURL) > 0 {
+			cmd.Data["ImageLink"] = imgURL
+		} else { // Try to upload
```

Comment:
What's the story behind trying to upload it? When would this happen?

Reply:
If the file is stored on local disk and not uploaded to an internet-accessible host. Then you add it as an email attachment.

Reply:
Oh I see